lib/Elevate/Components/CCS.pm

```diff
@@ -0,0 +1,424 @@
+                print $dh $txt;
+            }
+        }
+    }
+
+    return;
+}
+
+sub _dump_delegation_data_for_user ( $self, $user ) {
+    my $path = $self->_get_export_path_for_user($user);
+    my $dbh  = $self->_get_dbh();
+
+    my @webmail_users_info = Cpanel::CCS::Userdata::get_users($user);
+    my $delegates_ar       = Cpanel::CCS::Delegates::get( @webmail_users_info, $dbh );
+
+    my $delegate_file = $path . '/' . 'delegates.json';
+    Cpanel::JSON::DumpFile( $delegate_file, $delegates_ar );
+    return;
+}
+
+sub _dump_persistence_data_for_user ( $self, $user ) {
+    my $path             = $self->_get_export_path_for_user($user);
+    my $persistence_file = $path . '/' . 'persistence.json';
+    my $users_ccs_info   = $self->_get_ccs_info_for_user($user);
+
+    Cpanel::JSON::DumpFile( $persistence_file, $users_ccs_info );
+    return;
+}
+
+sub _make_backup_paths_for_user ( $self, $user ) {
+    my $path = $self->_get_export_path_for_user($user);
+    File::Path::make_path($path);
+
+    my %dump_types = DUMP_TYPES();
+    for ( keys(%dump_types) ) { File::Path::make_path("$path/$_"); }
+    return;
+}
+
+sub _get_query_for_type ( $self, $type, $uuid ) {
+    my %querydata = (
+        'calendars' => {
+            'args'  => [ $uuid, '1', 'f' ],
+            'query' => "SELECT icalendar_text
+                FROM
+                  calendar_object
+                  INNER JOIN calendar_bind ON calendar_bind.calendar_resource_id = calendar_object.calendar_resource_id
+                  INNER JOIN calendar_metadata ON calendar_metadata.resource_id = calendar_bind.calendar_resource_id
+                  INNER JOIN calendar_home ON calendar_home.resource_id = calendar_bind.calendar_home_resource_id
+                WHERE calendar_home.owner_uid = ?
+                AND calendar_bind.bind_status = ?
+                AND calendar_metadata.is_in_trash = ?;",
+        },
+        'contacts' => {
+            'args'  => [ $uuid, 'f' ],
+            'query' => "SELECT vcard_text
+                FROM
+                  addressbook_object
+                  INNER JOIN addressbook_home ON addressbook_home.resource_id = addressbook_object.addressbook_home_resource_id
+                WHERE addressbook_home.owner_uid = ?
+                AND addressbook_object.is_in_trash = ?;",
+        },
+    );
+
+    return ( $querydata{$type}{'query'}, $querydata{$type}{'args'} );
+}
+
+sub _get_dbh ($self) {
+    $self->{dbh} ||= Cpanel::CCS::DBUtils::get_dbh();
+    return $self->{dbh};
+}
+
+sub _get_export_path_for_user ( $self, $user ) {
+    $self->{$user}{export_path} ||= EXPORT_DIR() . '/' . $user . '/calendar_and_contacts';
+    return $self->{$user}{export_path};
+}
+
+sub _get_ccs_info_for_user ( $self, $user ) {
+    $self->{$user}{info} ||= Cpanel::CCS::Userdata::get_cpanel_account_users_uuids($user);
+    return $self->{$user}{info};
+}
+
+sub _ensure_export_directory ($self) {
+    File::Path::make_path(EXPORT_DIR);
```

Comment:
We need to make sure the perms on this path are 700.

Reply:
done